[MemoryLocation] Size Scalable Masked MemOps

MDevereau opened the pull request:
Scalable masked loads and stores with a get active lane mask whose size is less than or equal to the scalable minimum number of elements can be be proven to have a fixed size. Adding this infomation allows scalable masked loads and stores to benefit from alias analysis optimizations.

MDevereau committed:
[MemoryLocation] Size Scalable Masked MemOps

Scalable masked loads and stores with a get active lane mask whose size is less
than or equal to the scalable minimum number of elements can be be proven to
have a fixed size. Adding this infomation allows scalable masked loads and
stores to benefit from alias analysis optimizations.

nikic commented on llvm/lib/Analysis/MemoryLocation.cpp:
```diff
@@ -150,6 +150,29 @@ MemoryLocation::getForDest(const CallBase *CB, const TargetLibraryInfo &TLI) {
 
+static std::optional<FixedVectorType *>
+getFixedTypeFromScalableMemOp(Value *Mask, Type *Ty) {
+  auto ActiveLaneMask = dyn_cast<IntrinsicInst>(Mask);
```
```suggestion
  auto *ActiveLaneMask = dyn_cast<IntrinsicInst>(Mask);
```
etc

nikic commented on llvm/lib/Analysis/MemoryLocation.cpp:
```diff
@@ -150,6 +150,29 @@ MemoryLocation::getForDest(const CallBase *CB, const TargetLibraryInfo &TLI) {
+  if (!LaneMaskLo || !LaneMaskHi)
+    return std::nullopt;
+
+  uint64_t NumElts = LaneMaskHi->getZExtValue() - LaneMaskLo->getZExtValue();
```
Shouldn't this just use LaneMaskHi, without subtracting LaneMaskLo?

NumElts is the number of active elements, but the memory location is relative to the original base pointer, so inactive elements at the start need to be counted as well.

It looks like you are missing tests for a non-zero base index.

MDevereau replied:
> inactive elements at the start need to be counted as well.

From the langref for get active lane mask:

> %m[i] = icmp ult (%base + i), %n

From this definition I interpreted that it's not possible for the first N elements to be 0 and then have following 1s - either the mask returned from this begins with a 1, is all false, or poison. If there is a range of 3 between %base and %n, the new mask size would be 3. I added the test dead_scalar_store_offset for this scenario.

MDevereau committed:
Address review comments

Use patternmatch logic
Add pointer tokens to auto declarations
Add offset test dead_scalar_store_offset

david-arm reviewed (commented):
Thanks for addressing all the comments so far @MDevereau! I have a few more ...

MDevereau committed:
Add fixed-width tests

Use APInt maths
Refine tests

david-arm commented on llvm/lib/Analysis/MemoryLocation.cpp:
```diff
@@ -213,20 +241,26 @@ MemoryLocation MemoryLocation::getForArgument(const CallBase *Call,
-          LocationSize::upperBound(
-              DL.getTypeStoreSize(II->getArgOperand(0)->getType())),
-          AATags);
+          Arg, LocationSize::upperBound(DL.getTypeStoreSize(Ty)), AATags);
```
Not for this PR, but `upperBound` is currently extremely pessimistic for scalable vectors because for unknown pointers the upper bound is just the top of the address space. In future, we could improve the upper bound by taking vscale_range into account, or just add the ScalableBit to ImpreciseBit in the LocationSize object.

For normal stores we call this function instead:

```
  static LocationSize precise(TypeSize Value) {
    return LocationSize(Value.getKnownMinValue(), Value.isScalable());
  }
```

david-arm reviewed (approved):
LGTM!
